src/Ztm.WebApi/TransactionConfirmationWatcher.cs

```diff
@@ -0,0 +1,354 @@
+                {
+                    timers.Remove(transaction);
+                }
+            }
+            finally
+            {
+                timerLock.ExitWriteLock();
+            }
+        }
+
+        async Task<bool> StopTimer(uint256 transaction, Guid id)
+        {
+            timerLock.EnterWriteLock();
+            try
+            {
+                if (timers.TryGetValue(transaction, out var txTimers) && txTimers.TryGetValue(id, out var timer))
+                {
+                    await timer.Item1.StopAsync(CancellationToken.None);
+                    if (timer.Item1.ElapsedCount == 0)
+                    {
+                        RemoveTimer(transaction, id);
+                        return true;
+                    }
+                }
+            }
+            finally
+            {
+                timerLock.ExitWriteLock();
+            }
+
+            return false;
+        }
+
+        void ResumeTimer(ConfirmContext watch)
+        {
+            timerLock.EnterWriteLock();
+
+            try
+            {
+                TimerSetup(watch);
+            }
+            finally
+            {
+                timerLock.EnterWriteLock();
+            }
+        }
+
+        async Task Confirm(TransactionWatch<ConfirmContext> watch)
+        {
+            await Execute(watch.Context.Callback, watch.Context.Success, CancellationToken.None);
+        }
+
+        async Task Execute(Callback callback, TransactionConfirmationCallbackResult payload, CancellationToken cancellationToken)
+        {
+            await callbackRepository.AddHistoryAsync(callback.Id, payload, cancellationToken);
+
+            if (await this.callbackExecuter.Execute(callback.Id, callback.Url, payload, cancellationToken))
+            {
+                await callbackRepository.SetCompletedAsyc(callback.Id, cancellationToken);
+            }
+        }
+
+        // IBlockListener
+
+        public Task BlockAddedAsync(Block block, int height, CancellationToken cancellationToken)
+        {
+            return this.watcher.ExecuteAsync(block, height, BlockEventType.Added, cancellationToken);
+        }
+
+        public Task BlockRemovingAsync(Block block, int height, CancellationToken cancellationToken)
+        {
+            return this.watcher.ExecuteAsync(block, height, BlockEventType.Removing, cancellationToken);
+        }
+
+        // IHostedService
+
+        public Task StartAsync(CancellationToken cancellationToken)
+        {
+            return this.Initialize(cancellationToken);
+        }
+
+        public Task StopAsync(CancellationToken cancellationToken)
+        {
+            return this.StopAllTimers(cancellationToken);
+        }
+
+        // ITransactionConfirmationWatcherHandler
+
+        Task IWatcherHandler<TransactionWatch<ConfirmContext>, ConfirmContext>.AddWatchesAsync(IEnumerable<TransactionWatch<ConfirmContext>> watches, CancellationToken cancellationToken)
```

Comment:
We should group the implementation of `IWatcherHandler` together so it is easy to find.